integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/utils/PlanUtils3.java

```diff
@@ -81,14 +88,37 @@ public static void includeProviderFacet(
     Map<String, String> tableProperties = relation.table().properties();
 
     includeProviderFacet(tableCatalog, tableProperties, facets);
+
     Optional<DatasetIdentifier> datasetIdentifier =
         PlanUtils3.getDatasetIdentifier(context, tableCatalog, identifier, tableProperties);
 
-    if (datasetIdentifier.isPresent()) {
-      return Collections.singletonList(
-          datasetFactory.getDataset(datasetIdentifier.get(), relation.schema(), facets));
-    } else {
-      return Collections.emptyList();
+    return datasetIdentifier
+        .map(
+            value ->
+                Collections.singletonList(
+                    datasetFactory.getDataset(value, relation.schema(), facets)))
+        .orElse(Collections.emptyList());
+  }
+
+  public static Optional<DataSourceV2Relation> getDataSourceV2Relation(LogicalPlan x) {
```

Comment:
When moving code here, consider moving unit tests to another class.

Reply:
Unit tests in `TableContentChangeVisitorTest` still mostly test `TableContentChangeVisitor` I think. Unless you want to test `PlanUtils3.getDataSourceV2Relation` directly, but that I think doesn't make a lot of sense, this is just `instanceof` matching and test would only test ability to correctly write mock corresponding to mocked class 😉